### PR TITLE
Add Daffio api docs

### PR DESCRIPTION
**Pull request by lderrickable**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Daffio does not have api documentation for any of daffodil's packages.

## What is the new behavior?
Generated documentation for daffodil packages are now rendered in daffio.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
The actual documentation for the packages still needs to be written. This is just for the framework of rendering the generated documentation from dgeni.

### CODE REVIEW

**Comment by lderrickable** [comment]
This PR is blocked by https://github.com/graycoreio/daffodil/pull/349

**Comment by lderrickable** [comment]
> This PR is blocked by #349

#349 is merged

**Comment by xelaint** [comment]
Why is api-docs in daffio and not in design-land?

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss [hunk 1]
```diff
@@ -0,0 +1,128 @@
+@import 'utilities';
+@import "theme";
+
+.api-doc {
+  margin: 25px;
+
+
+  & a {
```
Comment (xelaint) [inline apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: not needed.

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    padding: 4px 8px;
+    text-transform: uppercase;
+    box-sizing: border-box;
+
+    @include breakpoint(mobile) {
+      margin: 0 0 0 15px;
+    }
+  }
+
+  &__label.class {
```
Comment (xelaint) [inline apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: All of this should just be `include @type-label();`

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  word-break: break-word;
+  
+  &__label {
+    font-size: 1rem;
+    line-height: 1em;
+    margin-right: 5px;
+  }
+}
+
+$border-color: 1px solid #ebebeb;
```
Comment (xelaint) [inline apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: `#ebebeb` is not a variable in the `$daff-gray` palette

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  margin-top: 15px;
+  box-sizing: border-box;
+
+  @include breakpoint(mobile) {
+    grid-template-columns: 1fr 3fr;
+  }
+
+  &__header {
+    display: flex;
+    background: #ebebeb;
```
Comment (xelaint) [inline apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: #ebebeb is not a variable in the $daff-gray palette, also background and border should not be the same color.

File: apps/daffio/src/scss/type-descriptors/_mixins.scss [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    background: daff-color($primary);
+  }
+
+  &.type-alias {
+    background: daff-color($secondary);
+    color: $white;
+  }
+
+  &.interface {
+    background: green;
```
Comment (xelaint) [inline apps/daffio/src/scss/type-descriptors/_mixins.scss]: `green` is not a color in `DaffPalette`.

File: apps/daffio/src/scss/type-descriptors/_mixins.scss [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    color: $white;
+  }
+
+  &.const {
+    background: purple;
```
Comment (xelaint) [inline apps/daffio/src/scss/type-descriptors/_mixins.scss]: `purple` is not a color in `DaffPalette`.

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss (hunk 4 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: changed the border color

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss (hunk 2 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: done

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss (hunk 1 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: removed

File: apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  doc: DaffioApiDoc = null;
+
+  constructor(
+    private docService: DaffioDocService,
+    private route: ActivatedRoute
+  ) { }
+
+  ngOnInit() {
+    // todo should navigate to 404 on a 404 error.
+    this.route.url.subscribe((url) => {
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts]: This needs test coverage.

File: apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      this.docService.getDoc(url[0] + '/' + url[1] + '/' + url[2]).subscribe(
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts]: I think this double subscribe should be replaced by a switch map. You also need to make sure you destroy the subscription, otherwise you leave a memory leak.

File: apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  docsList: DaffioApiDocReference[];
+
+  constructor(
+    private docService: DaffioDocService
+  ) {}
+
+  ngOnInit() {
+    this.docService.getDocsList().subscribe(
+      (data) => {
+        console.log(data[0].items);
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts]: Remove.

File: apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+   * A list of references for API documents.
+   */
+  docsList: DaffioApiDocReference[];
+
+  constructor(
+    private docService: DaffioDocService
+  ) {}
+
+  ngOnInit() {
+    this.docService.getDocsList().subscribe(
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts]: You need to cleanup the subscription, or use the async pipe in the template.

File: apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference-list.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { DaffioApiDocReference } from './api-doc-reference';
+
+/**
+ * A list of references for API documents.
+ */
+export interface DaffioApiDocReferenceList {
+  name: string;
+  title: string;
+  path: string;
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference-list.ts]: What is path in this case?

File: apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+/**
+ * An object for a reference to an API document.
+ */
+export interface DaffioApiDocReference {
+  name: string;
+  title: string;
+  path: string;
+  docType: string;
+  securityRisk: boolean;
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts]: Do we even use this?

File: apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+/**
+ * An object for a reference to an API document.
+ */
+export interface DaffioApiDocReference {
+  name: string;
+  title: string;
+  path: string;
+  docType: string;
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts]: Can this be more strongly typed?

File: apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+@Injectable({
+  providedIn: ApiDocServiceProviderModule
+})
+export class DaffioDocService {
+
+  constructor(private http: HttpClient) {}
+
+  getDoc(path: string): Observable<any> {
+    console.log(path);
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts]: Remove.

File: apps/daffio/src/app/content/docs/api-docs/views/api-doc/api-doc-view.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+describe('ApiDocViewComponent', () => {
+  let component: ApiDocViewComponent;
+  let fixture: ComponentFixture<ApiDocViewComponent>;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      imports: [
+        DaffioDocsApiModule,
+        RouterTestingModule,
+        NoopAnimationsModule
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/views/api-doc/api-doc-view.component.spec.ts]: Why do we need this NoopAnimationsModule?

File: apps/daffio/src/app/content/docs/api-docs/views/api-docs-list/api-docs-list-view.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+describe('ApiDocsListViewComponent', () => {
+  let component: ApiDocsListViewComponent;
+  let fixture: ComponentFixture<ApiDocsListViewComponent>;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      imports: [
+        DaffioDocsApiModule,
+        RouterTestingModule,
+        NoopAnimationsModule
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/views/api-docs-list/api-docs-list-view.component.spec.ts]: Why is this necessary?

File: apps/daffio/src/app/core/header/containers/header.component.ts [hunk 17]
```diff
@@ -10,6 +10,7 @@ import { ToggleSidebar } from '../../sidebar/actions/sidebar.actions';
 })
 export class DaffioHeaderContainer {
   links: any[] = [
+    {path: '/docs', title: 'Documentation'},
```
Comment (damienwebdev) [inline apps/daffio/src/app/core/header/containers/header.component.ts]: Let's just call it `Docs`

**Comment by damienwebdev** [comment]
@lderrickable I believe this also needs a rebase, as I think most of the dgeni work has been merged in a separate PR.

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+    box-sizing: border-box;
+
+    @include breakpoint(mobile) {
+      margin: 0 0 0 15px;
+    }
+  }
+  &__label {
+    @include type-label();
+  }
+}s
```
Comment (xelaint) [inline apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: random `s`

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+    margin: 15px 0 0 0;
+    padding: 4px 8px;
+    text-transform: uppercase;
+    box-sizing: border-box;
+
+    @include breakpoint(mobile) {
+      margin: 0 0 0 15px;
+    }
+  }
+  &__label {
```
Comment (xelaint) [inline apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: needs line spacer

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+  word-break: break-word;
+  
+  &__label {
+    font-size: 1rem;
+    line-height: 1em;
+    margin-right: 5px;
+  }
+}
+
+$border-color: 1px solid daff-color($daff-gray, 30);
```
Comment (xelaint) [inline apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: `20`

File: apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference-list.ts (hunk 11 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference-list.ts]: It is the path that shows in the url from `routerLink="{{doc.path}}"`. Then that path is taken by the api-doc/service and used as a directory lookup to retrieve the actual document.

File: apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts (hunk 12 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts]: No. Want me to remove it?

File: apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts (hunk 13 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts]: Like an enum or what?

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.spec.ts [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+    wrapper = fixture.componentInstance;
+    fixture.detectChanges();
+
+    component = fixture.debugElement.query(By.css('daffio-api-doc')).componentInstance;
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.spec.ts]: added

File: apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts (hunk 13 above)
Comment (damienwebdev) [reply apps/daffio/src/app/content/docs/api-docs/models/api-doc-reference.ts]: A simple `export type DaffApiDocType = "Component" | "Class"... ` would suffice.

File: apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+
+@Injectable({
+  providedIn: ApiDocServiceProviderModule
+})
+export class DaffioDocService {
+
+  constructor(private http: HttpClient) {}
+
+  getDoc(path: string): Observable<any> {
+    return this.http.get('/assets/daffio/' + path + '.json');
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts]: Move '/assets/daffio` into a class property.

File: apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+
+  constructor(
+    private docService: DaffioDocService
+  ) {}
+
+  ngOnInit() {
+    this.docsList$ = this.docService.getDocsList().pipe(
+      map(docsList => docsList[0].items),
+      catchError(error => {
+        //todo navigate to 404 page
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts]: Can we address this todo?

File: apps/daffio/src/app/content/docs/api-docs/components/api-docs-list/api-docs-list.component.html [hunk 24]
```diff
@@ -0,0 +1,9 @@
+<div class="api-docs-list">
+  <h1>API Index</h1>
+  <div class="api-docs-list__body">
+    <div class="list-item" *ngFor="let doc of docsList">
+      <label class="list-item__label {{ doc.docType }}">{{ doc.docType | slice:0:1 }}</label>
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-docs-list/api-docs-list.component.html]: Turned out the badge just had a white background, so I just needed to add the styles for the enum badge

File: apps/daffio/src/app/content/docs/api-docs/components/api-docs-list/api-docs-list.component.html [hunk 25]
```diff
@@ -0,0 +1,9 @@
...
+      <a class="list-item__name" routerLink="{{doc.path}}">{{ doc.title }}</a>
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-docs-list/api-docs-list.component.html]: added

File: apps/daffio/src/app/content/docs/api-docs/components/api-docs-list/api-docs-list.module.ts [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+import { RouterModule } from '@angular/router';
+
+import { DaffLinkModule } from '@daffodil/design';
+
+import { ApiDocsListComponent } from './api-docs-list.component';
+
+@NgModule({
+  imports: [
+    CommonModule,
+    DaffLinkModule,
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-docs-list/api-docs-list.module.ts]: we do now with the previous change request

File: apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts]: Made it better with `url.join('/')`

File: apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.scss [hunk 27]
```diff
@@ -0,0 +1,3 @@
+.docs-content {
+  box-sizing: border-box;
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.scss]: yep, removed the file

File: apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts (hunk 7 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts]: added

File: apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+  constructor(
+    private docService: DaffioDocService,
+    private route: ActivatedRoute
+  ) { }
+
+  ngOnInit() {
+    // todo should navigate to 404 on a 404 error.
+    this.doc$ = this.route.url.pipe(
+      switchMap((url) => {
+        return this.docService.getDoc(url.join('/'))
```
Comment (damienwebdev) [inline apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts]: Can we remove the nested pipe?

File: apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts (hunk 8 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts]: replaced with switchMap, and removed the subscription with an async pipe

File: apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts (hunk 9 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts]: removed

File: apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts (hunk 10 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts]: used an async pipe instead of subscription

File: apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts (hunk 14 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts]: removed

File: apps/daffio/src/app/content/docs/api-docs/views/api-doc/api-doc-view.component.spec.ts (hunk 15 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/views/api-doc/api-doc-view.component.spec.ts]: I think it was previously there to absorb the 404 routing errors being thrown by the `daff-api-doc-container` (or whatever it is), but I just mocked it out instead and removed the NoopAnimationsModule

File: apps/daffio/src/app/content/docs/api-docs/views/api-docs-list/api-docs-list-view.component.spec.ts (hunk 16 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/views/api-docs-list/api-docs-list-view.component.spec.ts]: See related comment

File: apps/daffio/src/app/core/header/containers/header.component.ts (hunk 17 above)
Comment (lderrickable) [reply apps/daffio/src/app/core/header/containers/header.component.ts]: Changed

**Comment by lderrickable** [comment]
> @lderrickable I believe this also needs a rebase, as I think most of the dgeni work has been merged in a separate PR.

yea, you right. Rebased

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss (hunk 18 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: removed

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss (hunk 19 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: what's a line spacer?

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss (hunk 20 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: changed

File: apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts (hunk 22 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts]: What about moving it into a CONST in the same file but outside of the class? Seems like it should be a const

File: apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts (hunk 23 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/containers/api-docs-list/api-docs-list.component.ts]: I'd prefer to do it in another PR.

File: apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts (hunk 28 above)
Comment (lderrickable) [reply apps/daffio/src/app/content/docs/api-docs/containers/api-doc/api-doc.component.ts]: Do you have an idea of how to do this?

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss (hunk 19 above)
Comment (damienwebdev) [reply apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: I think @xelaint means a new line between the braces.

File: apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts (hunk 22 above)
Comment (damienwebdev) [reply apps/daffio/src/app/content/docs/api-docs/services/api-doc.service.ts]: That would work.

File: apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss (hunk 19 above)
Comment (xelaint) [reply apps/daffio/src/app/content/docs/api-docs/components/api-doc/api-doc.component.scss]: yeah new line between the braces.

**Comment by damienwebdev** [comment]
closing in favor of #402
